src/taskgraph/transforms/notify.py

```diff
@@ -0,0 +1,174 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+"""
+Add notifications to tasks via Taskcluster's notify service.
+
+See https://docs.taskcluster.net/docs/reference/core/notify/usage for
+more information.
+"""
+from voluptuous import ALLOW_EXTRA, Any, Exclusive, Optional, Required
+
+from taskgraph.transforms.base import TransformSequence
+from taskgraph.util.schema import Schema, optionally_keyed_by, resolve_keyed_by
+
+_status_type = Any(
+    "on-completed",
+    "on-defined",
+    "on-exception",
+    "on-failed",
+    "on-pending",
+    "on-resolved",
+    "on-running",
+)
+
+_recipients = [
+    {
+        Required("type"): "email",
+        Required("address"): optionally_keyed_by("project", "level", str),
+        Optional("status-type"): _status_type,
+    },
+    {
+        Required("type"): "matrix-room",
+        Required("room-id"): str,
+        Optional("status-type"): _status_type,
+    },
+    {
+        Required("type"): "pulse",
+        Required("routing-key"): str,
+        Optional("status-type"): _status_type,
+    },
+    {
+        Required("type"): "slack-channel",
+        Required("channel-id"): str,
+        Optional("status-type"): _status_type,
+    },
+]
+
+_route_keys = {
+    "email": "address",
+    "matrix-room": "room-id",
+    "pulse": "routing-key",
+    "slack-channel": "channel-id",
+}
+"""Map each type to its primary key that will be used in the route."""
+
+NOTIFY_SCHEMA = Schema(
+    {
+        Exclusive("notify", "notify"): {
+            Required("recipients"): [Any(*_recipients)],
+            Optional("context"): {
+                Optional("email"): {
+                    Optional("subject"): str,
+                    Optional("content"): str,
+                    Optional("link"): {
+                        Required("text"): str,
+                        Required("href"): str,
+                    },
+                },
+                Optional("matrixBody"): str,
+                Optional("matrixFormattedBody"): str,
+                Optional("matrixFormat"): str,
+                Optional("matrixMsgType"): str,
```

Comment:
Bikeshed: I wonder if we should put the matrix-related attributes under a single dictionary, like the `email` one above. E.g.: 

```suggestion
                Optional("matrix"): {
                    Required("body"): str,
                    Optional("formatted-body"): str,
                    Optional("format"): str,
                    Optional("msg-type"): str,
                },
```

Reply:
Yes, this is a valid bikeshed and something I considered as well. However I just want to point out that these keys are set up to *exactly* match the Taskcluster's schema as per:
https://docs.taskcluster.net/docs/reference/core/notify/usage#setting-custom-messages

So the benefit of this way is that:
A) Users can lookup these keys in the Taskcluster docs easier, and
B) Taskgraph can just forward them wholesale to Taskcluster without doing any translation

I do agree that the Taskcluster schema is inconsistent and strange though (e.g why does `email` have subkeys, but not `matrix` or `slack`?).

I would be happy to make these requested changes if there is consensus this way is better. Just wanted to point out why I implemented them this way.

Reply:
Thanks for the context! I looked into the Taskcluster code and it seems nobody has ever raised this concern. The Matrix support was added in early 2020 in https://github.com/taskcluster/taskcluster/pull/2390. Slack was in late 2020 in https://github.com/taskcluster/taskcluster/pull/3610. As a reference point, email support was added in 2016[1].

People who implemented all 3 schemas are not at Mozilla anymore. So we won't be able to get any additional context.

I think it's okay to change the schema in Taskgraph for the sake of consistency. As a matter of fact, Taskgraph users are used to comply to some schema in `kind.yml` and let the transforms deal with what is really uploaded to the queue. That said, I'm happy to hear different opinions. 

[1] https://github.com/taskcluster/taskcluster-notify/commits/c557e553a703769dbe94629139865a100e4f1ca7/schemas/email-request.yml

Reply:
Ok those are fair points! Doing this means there would be no documentation, so I'll add a "How To" page for this as well.